File: db4oj/core/src/com/db4o/inside/freespace/FreespaceIxAddress.java
/* Copyright (C) 2004 - 2005  db4objects Inc.  http://www.db4o.com */

package com.db4o.inside.freespace;

import com.db4o.*;
import com.db4o.foundation.*;
import com.db4o.inside.*;
import com.db4o.inside.ix.*;


class FreespaceIxAddress extends FreespaceIx{
    
    FreespaceIxAddress(LocalObjectContainer file, MetaIndex metaIndex){
        super(file, metaIndex);
    }

    void add(int address, int length) {
        _index._handler.prepareComparison(new Integer(address));
        _indexTrans.add(length, new Integer(address));
    }

    int address() {
        return _visitor._value;
    }

    int length() {
        return _visitor._key;
    }

    void remove(int address, int length) {
        _index._handler.prepareComparison(new Integer(address));
        _indexTrans.remove(length, new Integer(address));
    }

    int freeSize(){
        final MutableInt mint = new MutableInt();
        final IntObjectVisitor freespaceVisitor = new IntObjectVisitor(){
            public void visit(int anInt, Object anObject) {
                mint.add(anInt);
            }
        };
        traverse(new Visitor4() {
            public void visit(Object obj) {
                ((IxTree)obj).visitAll(freespaceVisitor);
            }
        });
        return mint.value();
    }
}

File: db4on/core/src/core/com/db4o/internal/freespace/FreespaceIxAddress.cs
namespace com.db4o.@internal.freespace
{
	internal class FreespaceIxAddress : com.db4o.@internal.freespace.FreespaceIx
	{
		internal FreespaceIxAddress(com.db4o.@internal.LocalObjectContainer file, com.db4o.MetaIndex
			 metaIndex) : base(file, metaIndex)
		{
		}

		internal override void Add(int address, int length)
		{
			_index._handler.PrepareComparison(address);
			_indexTrans.Add(length, address);
		}

		internal override int Address()
		{
			return _visitor._value;
		}

		internal override int Length()
		{
			return _visitor._key;
		}

		internal override void Remove(int address, int length)
		{
			_index._handler.PrepareComparison(address);
			_indexTrans.Remove(length, address);
		}

		internal virtual int FreeSize()
		{
			com.db4o.foundation.MutableInt mint = new com.db4o.foundation.MutableInt();
			com.db4o.foundation.IntObjectVisitor freespaceVisitor = new _AnonymousInnerClass37
				(this, mint);
			Traverse(new _AnonymousInnerClass42(this, freespaceVisitor));
			return mint.Value();
		}

		private sealed class _AnonymousInnerClass37 : com.db4o.foundation.IntObjectVisitor
		{
			public _AnonymousInnerClass37(FreespaceIxAddress _enclosing, com.db4o.foundation.MutableInt
				 mint)
			{
				this._enclosing = _enclosing;
				this.mint = mint;
			}

			public void Visit(int anInt, object anObject)
			{
				mint.Add(anInt);
			}

			private readonly FreespaceIxAddress _enclosing;

			private readonly com.db4o.foundation.MutableInt mint;
		}

		private sealed class _AnonymousInnerClass42 : com.db4o.foundation.Visitor4
		{
			public _AnonymousInnerClass42(FreespaceIxAddress _enclosing, com.db4o.foundation.IntObjectVisitor
				 freespaceVisitor)
			{
				this._enclosing = _enclosing;
				this.freespaceVisitor = freespaceVisitor;
			}

			public void Visit(object obj)
			{
				((com.db4o.@internal.ix.IxTree)obj).VisitAll(freespaceVisitor);
			}

			private readonly FreespaceIxAddress _enclosing;

			private readonly com.db4o.foundation.IntObjectVisitor freespaceVisitor;
		}
	}
}
